src/test/libcephfs/vxattr.cc

```diff
@@ -0,0 +1,203 @@
+
+  // try setting a bad pool id
+  ASSERT_EQ(-EINVAL, ceph_setxattr(cmount, "test/d0", "ceph.dir.layout.pool_id", (void*)"300", 3, XATTR_REPLACE));
+
+  // try to set in invalid field
+  ASSERT_EQ(-ENODATA, ceph_setxattr(cmount, "test/d0", "ceph.dir.layout.bad_field", (void*)"300", 3, XATTR_REPLACE));
+
+  ASSERT_EQ(0, ceph_rmdir(cmount, "test/d0/subdir"));
+  ASSERT_EQ(0, ceph_rmdir(cmount, "test/d0"));
+  ASSERT_EQ(0, ceph_rmdir(cmount, "test"));
+
+  ceph_shutdown(cmount);
+}
+
+TEST(LibCephFS, GetAndSetDirPin) {
+
+  struct ceph_mount_info *cmount;
+  ASSERT_EQ(0, ceph_create(&cmount, NULL));
+  ASSERT_EQ(0, ceph_conf_read_file(cmount, NULL));
+  ASSERT_EQ(0, ceph_conf_parse_env(cmount, NULL));
+  ASSERT_EQ(0, ceph_mount(cmount, "/"));
+
+  ASSERT_EQ(0, ceph_mkdirs(cmount, "test/d1", 0777));
+
+  char value[1024];
+  int r = ceph_getxattr(cmount, "test/d1", "ceph.dir.pin", (void*)value, sizeof(value));
+  ASSERT_GT(r, 0);
+  value[r] = '\0';
+  ASSERT_STREQ("-1", value);
+
+  ASSERT_EQ(0, ceph_setxattr(cmount, "test/d1", "ceph.dir.pin", (void*)"1", 1, XATTR_CREATE));
+
+  r = ceph_getxattr(cmount, "test/d1", "ceph.dir.pin", (void*)value, sizeof(value));
+  ASSERT_GT(r, 0);
+  value[r] = '\0';
+  ASSERT_STREQ("1", value);
+
+  ASSERT_EQ(0, ceph_rmdir(cmount, "test/d1"));
+  ASSERT_EQ(0, ceph_rmdir(cmount, "test"));
+
+  ceph_shutdown(cmount);
+}
+
+TEST(LibCephFS, GetAndSetDirDistribution) {
+
+  struct ceph_mount_info *cmount;
+  ASSERT_EQ(0, ceph_create(&cmount, NULL));
+  ASSERT_EQ(0, ceph_conf_read_file(cmount, NULL));
+  ASSERT_EQ(0, ceph_conf_parse_env(cmount, NULL));
+  ASSERT_EQ(0, ceph_mount(cmount, "/"));
+
+  ASSERT_EQ(0, ceph_mkdirs(cmount, "test/d2", 0777));
+
+  char value[1024];
+  int r = ceph_getxattr(cmount, "test/d2", "ceph.dir.pin.distributed", (void*)value, sizeof(value));
+  ASSERT_GT(r, 0);
+  value[r] = '\0';
+  ASSERT_STREQ("0", value);
+
+  ASSERT_EQ(0, ceph_setxattr(cmount, "test/d2", "ceph.dir.pin.distributed", (void*)"1", 1, XATTR_CREATE));
+
+  r = ceph_getxattr(cmount, "test/d2", "ceph.dir.pin.distributed", (void*)value, sizeof(value));
+  ASSERT_GT(r, 0);
+  value[r] = '\0';
+  ASSERT_STREQ("1", value);
+
+  ASSERT_EQ(0, ceph_rmdir(cmount, "test/d2"));
+  ASSERT_EQ(0, ceph_rmdir(cmount, "test"));
+
+  ceph_shutdown(cmount);
+}
+
+TEST(LibCephFS, GetAndSetDirRandom) {
+
+  struct ceph_mount_info *cmount;
+  ASSERT_EQ(0, ceph_create(&cmount, NULL));
+  ASSERT_EQ(0, ceph_conf_read_file(cmount, NULL));
+  ASSERT_EQ(0, ceph_conf_parse_env(cmount, NULL));
+  ASSERT_EQ(0, ceph_mount(cmount, "/"));
+
+  ASSERT_EQ(0, ceph_mkdirs(cmount, "test/d3", 0777));
+
+  char value[1024];
+  int r = ceph_getxattr(cmount, "test/d3", "ceph.dir.pin.random", (void*)value, sizeof(value));
+  ASSERT_GT(r, 0);
+  value[r] = '\0';
+  ASSERT_STREQ("0", value);
+
+  ASSERT_EQ(0, ceph_setxattr(cmount, "test/d3", "ceph.dir.pin.random", (void*)"0.01", 4, XATTR_CREATE));
```

Comment:
recommend using a value that is exactly stored as a float, like 0.5.

Reply:
I didn't quite understand what you mean here.
Why is the string `"0.01"` not an acceptable value here ?
Values get passed as strings and the `(void*)` type cast is just to keep `ceph_setxattr()` happy.

Reply:
> Why is the string "0.01" not an acceptable value here ?

Because it's not exactly representable in base-2 floating point. Depending on the architecture and inner workings on the MDS, its printed representation may not be `"0.01"`.

s.f.: https://docs.oracle.com/cd/E19957-01/806-3568/ncg_goldberg.html

Reply:
ack; updated code to set value to `1.0/128.0`